tensorflow_addons/layers/multihead_attention.py

```diff
@@ -0,0 +1,253 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+import typing
+
+import tensorflow as tf
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class MultiHeadAttention(tf.keras.layers.Layer):
+    r"""
+    MultiHead Attention layer.
+
+    Defines the MultiHead Attention operation as defined in [Attention Is All You Need](https://arxiv.org/abs/1706.03762) which takes in a `query`, `key` and `value` tensors returns the dot-product attention between them:
+
+        ```python
+        mha = MultiHeadAttention(head_size=128, num_heads=128)
+
+        query = tf.random.uniform((32, 20, 200)) # (B, N, Dq)
+        key = tf.random.uniform((32, 15, 300)) # (B, M, Dk)
+        value = tf.random.uniform((32, 15, 400)) # (B, M, Dv)
+
+        attention = mha([query, key, value]) # (B, N, Dv)
+        ```
+
+    If `value` is not given then internally `value = key` will be used:
+
+         ```python
+        mha = MultiHeadAttention(head_size=128, num_heads=128)
+
+        query = tf.random.uniform((32, 20, 200)) # (B, N, Dq)
+        key = tf.random.uniform((32, 15, 300)) # (B, M, Dk)
+
+        attention = mha([query, key]) # (B, N, Dk)
+        ```
+
+    Arguments
+        head_size: int, dimensionality of the `query`, `key` and `value` tensors after the linear transformation.
+        num_heads: int, number of attention heads.
+        output_size: int, dimensionality of the output space, if `None` then the input dimension of `value` or `key` will be used, default `None`.
+        dropout_rate: float, `rate` parameter for the dropout layer that is applied to attention after softmax, default `0`.
```

Comment:
Let's remove the "rate" and call it dropout, which is consistent with all other keras layers.